Linkage of libIREECompiler.so.0 fails when IREE_TARGET_BACKEND_LLVM_CPU is not set

### What happened?

Trying to compile the sample `iree_samples_simple_embedding_simple_embedding_vmvx_sync` if `IREE_TARGET_BACKEND_DEFAULTS=OFF` and `IREE_TARGET_BACKEND_VMVX=ON` are set, the CMake target `iree_compiler_API_SharedImpl` fails with

```
[..]
 /usr/lib/x86_64-linux-gnu/libtinfo.so  third_party/iree/llvm-project/lib/libLLVMDemangle.a && :
/usr/bin/ld: cannot find -llldCOFF
/usr/bin/ld: cannot find -llldCommon
/usr/bin/ld: cannot find -llldELF
/usr/bin/ld: cannot find -llldMachO
/usr/bin/ld: cannot find -llldWasm
```

This can be resolved by setting `IREE_TARGET_BACKEND_LLVM_CPU=ON`. However, this increases the build time dramatically, with is essentially a not so nice thing for the [iree-template-cpp](https://github.com/iml130/iree-template-cpp) project in which it is demonstrated how to build a VMVX-based sample. This is also executed in a GitHub action (and this already takes quite long).

If it should not be legal to build the VMVX target backend without the LLVM CPU backend, CMake could already error on this.

### Steps to reproduce your issue

```bash
mkdir build
cd build
cmake -GNinja -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ -DIREE_HAL_DRIVER_DEFAULTS=OFF -DIREE_TARGET_BACKEND_DEFAULTS=OFF -DIREE_TARGET_BACKEND_VMVX=ON ..
cmake --build . --target iree_compiler_API_SharedImpl
```

### What component(s) does this issue relate to?

Compiler

### Version information

candidate-20230328.472

### Additional context

_No response_

--- a/compiler/src/iree/compiler/API/Internal/BUILD.bazel
+++ b/compiler/src/iree/compiler/API/Internal/BUILD.bazel
@@ -92,6 +92,9 @@ iree_compiler_cc_library(
     srcs = [
         "LLDToolEntryPoint.cpp",
     ],
+    # Special cased on the CMake side since we have to conditionally compile
+    # based on whether lld is enabled.
+    tags = ["skip-bazel_to_cmake"],
     deps = [
         "//compiler/bindings/c:headers",
         "@llvm-project//lld:COFF",

--- a/compiler/src/iree/compiler/API/Internal/CMakeLists.txt
+++ b/compiler/src/iree/compiler/API/Internal/CMakeLists.txt
@@ -115,10 +115,6 @@ if(IREE_LLD_TARGET)
   # Use generator expressions to conditionally include LLD targets, providing
   # a compiler definition if not available.
   macro(_add_dependent_lld_target target not_definition)
-    target_link_libraries(
-      iree_compiler_API_Internal_LLDToolEntryPoint
-      $<TARGET_NAME_IF_EXISTS:${target}>
-    )
     list(APPEND _lld_deps
       $<TARGET_NAME_IF_EXISTS:${target}>)
     list(APPEND _lld_copts

--- a/compiler/src/iree/compiler/API/Internal/CMakeLists.txt
+++ b/compiler/src/iree/compiler/API/Internal/CMakeLists.txt
@@ -86,23 +86,6 @@ iree_cc_library(
 
 iree_cc_library(
   NAME
-    LLDToolEntryPoint
-  SRCS
-    "LLDToolEntryPoint.cpp"
-  DEPS
-    LLVMSupport
-    LLVMTargetParser
-    iree::compiler::bindings::c::headers
-    lldCOFF
-    lldCommon
-    lldELF
-    lldMachO
-    lldWasm
-  PUBLIC
-)
-
-iree_cc_library(
-  NAME
     MLIRInterop
   SRCS
     "MLIRInterop.cpp"
@@ -124,3 +107,43 @@ iree_cc_library(
 )
 
 ### BAZEL_TO_CMAKE_PRESERVES_ALL_CONTENT_BELOW_THIS_LINE ###
+
+set(_lld_copts)
+set(_lld_deps)
+
+if(IREE_LLD_TARGET)
+  # Use generator expressions to conditionally include LLD targets, providing
+  # a compiler definition if not available.
+  macro(_add_dependent_lld_target target not_definition)
+    target_link_libraries(
+      iree_compiler_API_Internal_LLDToolEntryPoint
+      $<TARGET_NAME_IF_EXISTS:${target}>
+    )
+    list(APPEND _lld_deps
+      $<TARGET_NAME_IF_EXISTS:${target}>)
+    list(APPEND _lld_copts
+      $<IF:$<TARGET_EXISTS:${target}>,,-D${not_definition}>)
+  endmacro()
+  list(APPEND _lld_deps lldCommon)
+  _add_dependent_lld_target(lldCOFF IREE_COMPILER_LLD_COFF_DISABLED)
+  _add_dependent_lld_target(lldELF IREE_COMPILER_LLD_ELF_DISABLED)
+  _add_dependent_lld_target(lldMachO IREE_COMPILER_LLD_MACHO_DISABLED)
+  _add_dependent_lld_target(lldWasm IREE_COMPILER_LLD_WASM_DISABLED)
+else()
+  list(APPEND _lld_copts -DIREE_COMPILER_LLD_DISABLED)
+endif()
+
+iree_cc_library(
+  NAME
+    LLDToolEntryPoint
+  SRCS
+    "LLDToolEntryPoint.cpp"
+  COPTS
+    ${_lld_copts}
+  DEPS
+    LLVMSupport
+    LLVMTargetParser
+    iree::compiler::bindings::c::headers
+    ${_lld_deps}
+  PUBLIC
+)

--- a/compiler/src/iree/compiler/API/Internal/LLDToolEntryPoint.cpp
+++ b/compiler/src/iree/compiler/API/Internal/LLDToolEntryPoint.cpp
@@ -9,6 +9,16 @@
 // duplicate the flavor parsing and invoke the backend directly similar to
 // what the lldMain() does.
 
+#ifdef IREE_COMPILER_LLD_DISABLED
+#include <stdio.h>
+
+#include "iree/compiler/tool_entry_points_api.h"
+
+int ireeCompilerRunLldMain(int argc, char **argv) {
+  fprintf(stderr, "IREE was not built with support for LLD\n");
+  return 1;
+}
+#else
 #include <cstdlib>
 #include <vector>
 
@@ -82,17 +92,34 @@ int ireeCompilerRunLldMain(int argc, char **argv) {
   std::vector<const char *> args(argv, argv + argc);
   switch (parseFlavor(args)) {
     case Gnu:
+#ifndef IREE_COMPILER_LLD_ELF_DISABLED
       return !elf::link(args, stdoutOS, stderrOS, exitEarly, disableOutput);
+#else
+      die("lld is not compiled with ELF support");
+#endif
     case WinLink:
+#ifndef IREE_COMPILER_LLD_COFF_DISABLED
       return !coff::link(args, stdoutOS, stderrOS, exitEarly, disableOutput);
+#else
+      die("lld is not compiled with COFF support");
+#endif
     case Darwin:
+#ifndef IREE_COMPILER_LLD_MACHO_DISABLED
       return !macho::link(args, stdoutOS, stderrOS, exitEarly, disableOutput);
+#else
+      die("lld is not compiled with MachO support");
+#endif
     case Wasm:
+#ifndef IREE_COMPILER_LLD_WASM_DISABLED
       return !lld::wasm::link(args, stdoutOS, stderrOS, exitEarly,
                               disableOutput);
+#else
+      die("lld is not compiled with WASM support");
+#endif
     default:
       die("lld is a generic driver.\n"
           "Invoke ld.lld (Unix), ld64.lld (macOS), lld-link (Windows), wasm-ld"
           " (WebAssembly) instead");
   }
 }
+#endif  // IREE_COMPILER_LLD_DISABLED

